File: frontend/src/lib/validation/phone.ts
import { z } from "zod";

/**
 * Vietnam mobile numbers — the only phones this bakery takes, since pickup
 * contact means a phone that can receive a call or a message.
 *
 * Stored as E.164 (`+84912345678`), entered after a fixed `+84` as the nine-digit
 * national number, shown as `+84 912 345 678`. The backend applies the same rule
 * (`internal/shared/utils/phone.go`); both are tested against
 * `contracts/vietnam-phone-cases.json`.
 */

/**
 * The national number: a carrier prefix MIC has assigned, then seven digits.
 * 095 is retired; the 11-digit 01x numbers moved to these prefixes in 2018.
 * When MIC assigns a new prefix, add it here, in the backend and in the cases.
 */
const MOBILE_NUMBER = /^(?:3[2-9]|5[25689]|7[06-9]|8[1-9]|9[0-46-9])\d{7}$/;

/** Digits in a national mobile number: a two-digit prefix and seven more. */
export const NATIONAL_NUMBER_LENGTH = 9;

/** Digits and the separators people type between them; anything else is not a phone. */
const PHONE_CHARACTERS = /^[\d\s.()+-]*$/;

/**
 * The national number inside whatever was typed or pasted. `00` is Vietnam's
 * international access code and `84` the country code, so both come off the
 * front of a long number; then the trunk `0` does, since no national number
 * starts with 0. Zeros inside the number are never touched.
 */
export function nationalNumber(raw: string): string {
  let digits = raw.replace(/\D/g, "");
  if (digits.startsWith("00")) {
    digits = digits.slice(2);
  }
  if (digits.startsWith("84") && digits.length > 9) {
    digits = digits.slice(2);
  }
  return digits.replace(/^0+/, "");
}

/** `912345678` → `912 345 678`: groups of three, the way the number is read aloud. */
export function formatNationalNumber(national: string): string {
  return national.replace(/(\d{3})(?=\d)/g, "$1 ");
}

/** The stored form, or null when the input is not a Vietnam mobile number. */
export function normalizeVietnamPhone(raw: string): string | null {
  if (!PHONE_CHARACTERS.test(raw)) {
    return null;
  }
  const national = nationalNumber(raw);
  return MOBILE_NUMBER.test(national) ? `+84${national}` : null;
}

/**
 * `+84912345678` → `+84 912 345 678`. A value this cannot read is handed back
 * untouched: rows written before the rule existed still have to render.
 */
export function formatVietnamPhone(stored: string | null | undefined): string {
  if (!stored) {
    return "";
  }
  const normalized = normalizeVietnamPhone(stored);
  if (!normalized) {
    return stored;
  }
  return `+84 ${formatNationalNumber(normalized.slice(3))}`;
}

export const PHONE_FORMAT_MESSAGE = "Enter a Vietnam mobile number";

export const PHONE_TAKEN_MESSAGE = "This number is already on another account";

/**
 * The phone field of every write schema. The form holds the national number;
 * the API receives E.164. An empty field is sent as "" — a PATCH clears the
 * stored phone, a create stores none.
 */
export function vietnamPhoneZodString() {
  return z
    .string()
    .trim()
    .transform((value, ctx) => {
      if (value === "") {
        return "";
      }
      const stored = normalizeVietnamPhone(value);
      if (stored === null) {
        ctx.addIssue({ code: z.ZodIssueCode.custom, message: PHONE_FORMAT_MESSAGE });
        return z.NEVER;
      }
      return stored;
    })
    .optional();
}

File: backend/internal/shared/utils/phone.go
package utils

import (
	"regexp"
	"strings"
)

// Vietnam numbers in the spellings people type: national (leading zero) and
// international (+84 or 84, sometimes followed by the national zero). After the
// prefix comes either a mobile number — 3, 5, 7, 8 or 9 then eight digits — or a
// land line: 2, the rest of the area code and the subscriber number, ten digits
// in all. Nothing has been assigned to 0, 1, 4 or 6 since the 2018 renumber.
// The frontend applies the same pattern (src/lib/validation/phone.ts); both are
// tested against contracts/vietnam-phone-cases.json.
var (
	vietnamPhonePattern  = regexp.MustCompile(`^(?:0|\+?840?)(2\d{9}|[35789]\d{8})$`)
	phoneSeparatorsRegex = regexp.MustCompile(`[\s.()\-]`)
)

// NormalizeVietnamPhone returns the stored E.164 form (+84…) of a Vietnam phone
// number and reports whether the input was one. Separators are how people type,
// not what is stored, so they are dropped before matching.
func NormalizeVietnamPhone(phone string) (string, bool) {
	m := vietnamPhonePattern.FindStringSubmatch(phoneSeparatorsRegex.ReplaceAllString(phone, ""))
	if m == nil {
		return "", false
	}
	return "+84" + m[1], true
}

// NormalizeVietnamPhonePtr normalizes an optional phone from a request. nil stays
// nil (absent) and a blank value becomes "" (a request to clear); anything else
// must be a Vietnam number, and ok is false when it is not.
func NormalizeVietnamPhonePtr(phone *string) (normalized *string, ok bool) {
	if phone == nil {
		return nil, true
	}
	trimmed := strings.TrimSpace(*phone)
	if trimmed == "" {
		return &trimmed, true
	}
	e164, ok := NormalizeVietnamPhone(trimmed)
	if !ok {
		return nil, false
	}
	return &e164, true
}
